filesystem/File.php

```diff
@@ -686,114 +593,117 @@ protected function updateLinks($old, $new) {
 	public function getAbsoluteURL() {
-		return Director::absoluteURL($this->getURL());
+		$url = $this->getURL();
+		if($url) {
+			return Director::absoluteURL($url);
+		}
 	}
 
 	/**
-	 * Gets the relative URL accessible through the web.
+	 * Gets the URL of this file
 	 *
 	 * @uses Director::baseURL()
 	 * @return string
 	 */
 	public function getURL() {
-		return Controller::join_links(Director::baseURL(), $this->getFilename());
-	}
-
-	/**
-	 * Returns an absolute filesystem path to the file.
-	 * Use {@link getRelativePath()} to get the same path relative to the webroot.
-	 *
-	 * @return String
-	 */
-	public function getFullPath() {
-		$baseFolder = Director::baseFolder();
-
-		if(strpos($this->getFilename(), $baseFolder) === 0) {
-			// if path is absolute already, just return
-			return $this->getFilename();
-		} else {
-			// otherwise assume silverstripe-basefolder
-			return Director::baseFolder() . '/' . $this->getFilename();
+		if($this->File->exists()) {
+			return $this->File->getURL();
 		}
 	}
 
 	/**
-	 * Returns path relative to webroot.
-	 * Serves as a "fallback" method to create the "Filename" property if it isn't set.
-	 * If no {@link Folder} is set ("ParentID" property),
-	 * defaults to a filename relative to the ASSETS_DIR (usually "assets/").
+	 * Get URL, but without resampling.
 	 *
-	 * @return String
+	 * @return string
 	 */
-	public function getRelativePath() {
-		if($this->ParentID) {
-			// Don't use the cache, the parent has just been changed
-			$p = DataObject::get_by_id('Folder', $this->ParentID, false);
-			if($p && $p->exists()) return $p->getRelativePath() . $this->getField("Name");
-			else return ASSETS_DIR . "/" . $this->getField("Name");
-		} else if($this->getField("Name")) {
-			return ASSETS_DIR . "/" . $this->getField("Name");
-		} else {
-			return ASSETS_DIR;
+	public function getSourceURL() {
+		if($this->File->exists()) {
+			return $this->File->getSourceURL();
 		}
 	}
 
 	/**
 	 * @todo Coupling with cms module, remove this method.
+	 *
+	 * @return string
 	 */
 	public function DeleteLink() {
 		return Director::absoluteBaseURL()."admin/assets/removefile/".$this->ID;
 	}
 
+	/**
+	 * Get Filename for the currently stored underlying asset
+	 *
+	 * @return string
+	 */
 	public function getFilename() {
-		// Default behaviour: Return field if its set
-		if($this->getField('Filename')) {
-			return $this->getField('Filename');
-		} else {
-			return ASSETS_DIR . '/';
-		}
+		return $this->File->getFilename();
 	}
 
 	/**
-	 * Caution: this does not change the location of the file on the filesystem.
+	 * Get the expected value for "Filename" once this record is saved, and the underlying
+	 * file is moved.
+	 *
+	 * It is necessary for this to be distinct from getFilename in order to ensure that
+	 * we can predict its value prior to save.
+	 *
+	 * @return string
 	 */
-	public function setFilename($val) {
-		$this->setField('Filename', $val);
+	public function getExpectedFilename() {
```

Comment:
I'll make this getFilename and remove the other method. Internal code will have to invoke $file->File->getFilename() to get the internal value.